txn/buffer.go

```diff
@@ -0,0 +1,358 @@
+LOOP:
+	for _, op := range state.buffer {
+		select {
+		case opChan <- op:
+		case <-state.stopChan:
+			errChan <- ErrTxnAborted
+			break LOOP
+		}
+	}
+	close(opChan)
+	close(errChan)
+	state.wg.Done()
+	b.wg.Done()
+}
+
+// PurgeTxn closes any transaction streams in progress and deletes all oplog
+// entries associated with a transaction.
+//
+// Must not be called concurrently with other transaction-related operations.
+// For a given transaction, it must not be called until after a final oplog
+// entry has been passed to AddOp and it must not be called more than once.
+func (b *Buffer) PurgeTxn(m Meta) error {
+	b.Lock()
+	defer b.Unlock()
+	if b.stopped {
+		return ErrBufferClosed
+	}
+	state := b.txns[m.id]
+	if state == nil {
+		return fmt.Errorf("PurgeTxn found no state for %v", m.id)
+	}
+
+	// When the lock is dropped, we don't want Stop to find this transaction and
+	// double-close it.
+	delete(b.txns, m.id)
+	close(state.stopChan)
+
+	// Wait for goroutines to terminate, then clean up.
+	state.wg.Wait()
+	state.purge()
+
+	return nil
+}
+
+// Stop shuts down processing and cleans up.  Subsequent calls to Stop() will return nil.
+// All other methods error after this is called.
+func (b *Buffer) Stop() error {
+	b.Lock()
+	if b.stopped {
+		b.Unlock()
+		return nil
+	}
+
+	b.stopped = true
+	for _, state := range b.txns {
+		close(state.stopChan)
+	}
+
+	b.Unlock()
+
+	// At this point we know any subsequent public method will see the buffer
+	// is stopped, no new goroutines will be launched, and existing goroutines
+	// have been signaled to close.  Next, wait for goroutines to stop, then
+	// clean up.
+
+	b.wg.Wait()
+	errs := make([]error, 0)
+	for _, state := range b.txns {
+		err := state.purge()
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) != 0 {
+		return errs[0]
+	}
+
+	return nil
+}
+
+// sendErrAndClose is a utility for putting an error on a channel before closing.
+func sendErrAndClose(o chan db.Oplog, e chan error, err error) (chan db.Oplog, chan error) {
+	e <- err
+	close(o)
+	close(e)
+	return o, e
+}
+
+const extractErrorFmt = "error extracting transaction ops: %s: %v"
+
+func extractInnerOps(doc bson.D) ([]db.Oplog, error) {
+	rawAO, err := bsonutil.FindValueByKey("applyOps", &doc)
+	if err != nil {
+		return nil, fmt.Errorf(extractErrorFmt, "applyOps field", err)
+	}
+
+	ao, ok := rawAO.(bson.A)
+	if !ok {
+		return nil, fmt.Errorf(extractErrorFmt, "applyOps field", "not a BSON array")
+	}
+
+	ops := make([]db.Oplog, len(ao))
+	for i, v := range ao {
+		opDoc, ok := v.(bson.D)
+		if !ok {
+			return nil, fmt.Errorf(extractErrorFmt, "applyOps op", "not a BSON document")
+		}
+		op, err := bsonDocToOplog(opDoc)
+		if err != nil {
+			return nil, fmt.Errorf(extractErrorFmt, "applyOps op", err)
+		}
+		ops[i] = *op
+	}
+
+	return ops, nil
+}
+
+const opConvertErrorFmt = "error converting bson.D to op: %s: %v"
+
+func bsonDocToOplog(doc bson.D) (*db.Oplog, error) {
+	op := db.Oplog{}
```

Comment:
nit: stylistically, I prefer using the zero-value initializers for non-pointer types if there's no special logic needed (i.e. `var op db.Oplog`)